Library/OcMainLib/OpenCoreUefiUnloadDrivers.c

```diff
@@ -0,0 +1,352 @@
+
+  Status = gBS->OpenProtocol (
+                  TheHandle,
+                  &gEfiComponentName2ProtocolGuid,
+                  (VOID **)&CompNameStruct,
+                  gImageHandle,
+                  NULL,
+                  EFI_OPEN_PROTOCOL_GET_PROTOCOL
+                  );
+  if (!EFI_ERROR (Status)) {
+    BestLang = GetBestLanguageForDriver (CompNameStruct->SupportedLanguages, Language, FALSE);
+    Status   = CompNameStruct->GetDriverName (CompNameStruct, BestLang, &RetVal);
+    if (BestLang != NULL) {
+      FreePool (BestLang);
+      BestLang = NULL;
+    }
+
+    if (!EFI_ERROR (Status)) {
+      return (RetVal);
+    }
+  }
+
+  Status = gBS->OpenProtocol (
+                  TheHandle,
+                  &gEfiComponentNameProtocolGuid,
+                  (VOID **)&CompNameStruct,
+                  gImageHandle,
+                  NULL,
+                  EFI_OPEN_PROTOCOL_GET_PROTOCOL
+                  );
+  if (!EFI_ERROR (Status)) {
+    BestLang = GetBestLanguageForDriver (CompNameStruct->SupportedLanguages, Language, FALSE);
+    Status   = CompNameStruct->GetDriverName (CompNameStruct, BestLang, &RetVal);
+    if (BestLang != NULL) {
+      FreePool (BestLang);
+    }
+
+    if (!EFI_ERROR (Status)) {
+      return (RetVal);
+    }
+  }
+
+  return (NULL);
+}
+
+VOID
+ReportImageName (
+  IN VOID          *Context,
+  IN CONST CHAR16  *Name,
+  IN EFI_HANDLE    Handle
+  )
+{
+  DRIVER_REPORT_CONTEXT  *ReportContext;
+
+  ReportContext = Context;
+
+  OcAsciiPrintBuffer (
+    &ReportContext->FileBuffer,
+    &ReportContext->FileBufferSize,
+    "%s\n",
+    Name
+    );
+}
+
+VOID
+UnloadImageByName (
+  IN VOID          *Context,
+  IN CONST CHAR16  *Name,
+  IN EFI_HANDLE    Handle
+  )
+{
+  EFI_STATUS            Status;
+  UINTN                 Index;
+  UNLOAD_IMAGE_CONTEXT  *UnloadContext;
+
+  UnloadContext = Context;
+
+  for (Index = 0; Index < UnloadContext->UnloadNameCount; Index++) {
+    if (StrCmp (UnloadContext->UnloadInfo[Index].UnloadName, Name) == 0) {
+      Status = gBS->UnloadImage (Handle);
+      DEBUG ((
+        EFI_ERROR (Status) ? DEBUG_WARN : DEBUG_INFO,
+        "OC: UnloadImage %s - %r\n",
+        UnloadContext->UnloadInfo[Index].UnloadName,
+        Status
+        ));
+      UnloadContext->UnloadInfo[Index].Unloaded = TRUE; ///< Do not report as 'not found' even if we failed to unload it.
+      break;
+    }
+  }
+}
+
+//
+// Perform action for all driver binding protocols which are
+// present on the same handle as loaded image protocol and
+// where the handle has a usable name.
+//
+// Note: In terms of choosing a name to identify the driver, while the loaded image
+// section name is shorter and corresponds better to the driver's file name, and the
+// firmware volume GUID from the loaded image path identifies the driver more uniquely,
+// these are both not always available for a loaded image (across various firmware),
+// whereas the driver component name normally is.
```

Comment:
This is one of two main choice points within this PR: there are various other names that would seem preferable (more compact, arguably more definitive) that the long English language (usually...) component names. But the other names are often not available; component name seems to be widely available (i.e. in almost all firmware).

Reply:
Yes, I don’t think there is much else you can do. The “proper” names are FFS metadata not used for loading iirc.